File: cplusplus/sssp.cpp
#include "sssp.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>
#include <unordered_map>
#include <unordered_set>

namespace sssp {

namespace {

constexpr size_t kMaxIter = 100000000;
constexpr double kEps = 1e-12;
constexpr double kInf = std::numeric_limits<double>::infinity();

size_t Pow2(size_t exp) {
  if (exp == 0) return 1;
  return size_t(1) << std::min(exp, size_t(30));
}

void Relax(std::vector<double>& d, std::vector<std::optional<size_t>>& pred,
           size_t u, size_t v, double w) {
  double new_d = d[u] + w;
  if (new_d > d[v]) return;
  d[v] = new_d;
  pred[v] = u;
}

int F64Cmp(double a, double b) {
  if (a < b) return -1;
  if (a > b) return 1;
  return 0;
}

// --- MinHeap ---
class MinHeap {
 public:
  explicit MinHeap(size_t max_vertex)
      : index_(max_vertex, static_cast<size_t>(-1)) {}

  bool is_empty() const { return heap_.empty(); }
  bool contains(size_t v) const {
    return v < index_.size() && index_[v] != static_cast<size_t>(-1);
  }

  void insert(size_t v, double dist) {
    size_t i = heap_.size();
    index_[v] = i;
    heap_.emplace_back(v, dist);
    sift_up(i);
  }

  std::pair<size_t, double> extract_min() {
    auto top = heap_.front();
    index_[top.first] = static_cast<size_t>(-1);
    heap_.front() = heap_.back();
    heap_.pop_back();
    if (!heap_.empty()) {
      index_[heap_.front().first] = 0;
      sift_down(0);
    }
    return top;
  }

  void decrease_key(size_t v, double new_d) {
    size_t i = index_[v];
    if (i == static_cast<size_t>(-1) || heap_[i].second <= new_d) return;
    heap_[i] = {v, new_d};
    sift_up(i);
  }

 private:
  std::vector<std::pair<size_t, double>> heap_;
  std::vector<size_t> index_;

  void sift_up(size_t i) {
    while (i > 0) {
      size_t p = (i - 1) / 2;
      if (heap_[p].second <= heap_[i].second) break;
      swap(i, p);
      i = p;
    }
  }

  void sift_down(size_t i) {
    while (true) {
      size_t l = 2 * i + 1, r = 2 * i + 2, small = i;
      if (l < heap_.size() && heap_[l].second < heap_[small].second) small = l;
      if (r < heap_.size() && heap_[r].second < heap_[small].second) small = r;
      if (small == i) break;
      swap(i, small);
      i = small;
    }
  }

  void swap(size_t i, size_t j) {
    std::pair<size_t, double> a = heap_[i], b = heap_[j];
    heap_[i] = b;
    heap_[j] = a;
    index_[a.first] = j;
    index_[b.first] = i;
  }
};

// --- FrontierDS ---
class FrontierDS {
 public:
  FrontierDS(size_t m, double b) : m_(std::max(m, size_t(1))), b_(b) {}

  void insert(size_t key, double value) {
    auto it = key_to_value_.find(key);
    if (it != key_to_value_.end()) {
      if (value >= it->second) return;
      list_.erase(std::remove_if(list_.begin(), list_.end(),
                                 [key](const auto& p) { return p.second == key; }),
                  list_.end());
    }
    key_to_value_[key] = value;
    list_.emplace_back(value, key);
    sorted_ = false;
  }

  void batch_prepend(const std::vector<std::pair<size_t, double>>& pairs) {
    for (const auto& [key, value] : pairs) {
      auto it = key_to_value_.find(key);
      if (it != key_to_value_.end() && value >= it->second) continue;
      list_.erase(std::remove_if(list_.begin(), list_.end(),
                                 [key](const auto& p) { return p.second == key; }),
                  list_.end());
      key_to_value_[key] = value;
      list_.emplace_back(value, key);
    }
    sorted_ = false;
  }

  bool pull(double& bound, std::vector<size_t>& keys) {
    if (list_.empty()) {
      bound = b_;
      return false;
    }
    if (!sorted_) {
      std::sort(list_.begin(), list_.end(), [](const auto& a, const auto& b) {
        int c = F64Cmp(a.first, b.first);
        return c != 0 ? c < 0 : a.second < b.second;
      });
      sorted_ = true;
    }
    size_t take = std::min(m_, list_.size());
    keys.clear();
    keys.reserve(take);
    for (size_t i = 0; i < take; ++i) {
      keys.push_back(list_[i].second);
      key_to_value_.erase(list_[i].second);
    }
    list_.erase(list_.begin(), list_.begin() + static_cast<std::ptrdiff_t>(take));
    bound = list_.empty() ? b_ : list_.front().first;
    return true;
  }

  bool is_empty() const { return list_.empty(); }

 private:
  size_t m_;
  double b_;
  std::unordered_map<size_t, double> key_to_value_;
  std::vector<std::pair<double, size_t>> list_;
  bool sorted_ = true;
};

static size_t count_subtree(size_t u, const std::vector<std::vector<size_t>>& children,
                            std::vector<size_t>& size) {
  if (size[u] != 0) return size[u];
  size_t s = 1;
  for (size_t v : children[u]) s += count_subtree(v, children, size);
  size[u] = s;
  return s;
}

std::pair<std::vector<size_t>, std::vector<size_t>> find_pivots(
    const Graph& g, std::vector<double>& d,
    std::vector<std::optional<size_t>>& pred, double b,
    const std::vector<size_t>& s, size_t k) {
  std::vector<size_t> w = s, wi = s;

  for (size_t round = 1; round <= k; ++round) {
    std::vector<size_t> wi_next;
    for (size_t u : wi) {
      for (const auto& [v, w_e] : g.out_edges(u)) {
        double new_d = d[u] + w_e;
        if (new_d > d[v]) continue;
        Relax(d, pred, u, v, w_e);
        if (new_d < b) wi_next.push_back(v);
      }
    }
    for (size_t v : wi_next) w.push_back(v);
    wi = std::move(wi_next);
    if (w.size() > k * s.size()) return {s, w};
  }

  std::unordered_set<size_t> in_w(w.begin(), w.end());
  std::vector<std::optional<size_t>> parent(g.vertex_count());
  for (size_t u : w) {
    for (const auto& [v, w_e] : g.out_edges(u)) {
      if (in_w.count(v) && std::fabs(d[v] - (d[u] + w_e)) < kEps) {
        if (!parent[v]) parent[v] = u;
      }
    }
  }

  std::vector<std::vector<size_t>> children(g.vertex_count());
  for (size_t v : w) {
    if (parent[v]) children[*parent[v]].push_back(v);
  }

  std::vector<size_t> subtree_size(g.vertex_count(), 0);
  std::unordered_set<size_t> has_parent;
  for (size_t v : w)
    if (parent[v]) has_parent.insert(v);

  std::vector<size_t> roots_in_s;
  for (size_t r : s) {
    if (!has_parent.count(r)) {
      if (count_subtree(r, children, subtree_size) >= k) roots_in_s.push_back(r);
    }
  }
  return {roots_in_s, w};
}

std::pair<double, std::vector<size_t>> base_case(
    const Graph& g, std::vector<double>& d,
    std::vector<std::optional<size_t>>& pred, double b,
    const std::vector<size_t>& s, size_t n, size_t k) {
  size_t x = s[0];
  std::vector<size_t> u0;
  MinHeap heap(n);
  heap.insert(x, d[x]);

  while (!heap.is_empty() && u0.size() < k + 1) {
    auto [u, du] = heap.extract_min();
    u0.push_back(u);
    for (const auto& [v, w_e] : g.out_edges(u)) {
      double new_d = du + w_e;
      if (new_d >= b || new_d > d[v]) continue;
      Relax(d, pred, u, v, w_e);
      if (heap.contains(v))
        heap.decrease_key(v, d[v]);
      else
        heap.insert(v, d[v]);
    }
  }

  if (u0.size() <= k) return {b, u0};
  double b_prime = b;
  for (size_t v : u0)
    if (d[v] > b_prime) b_prime = d[v];
  std::vector<size_t> filtered;
  for (size_t v : u0)
    if (d[v] < b_prime) filtered.push_back(v);
  return {b_prime, filtered};
}

std::pair<double, std::vector<size_t>> bmssp(
    const Graph& g, std::vector<double>& d,
    std::vector<std::optional<size_t>>& pred, size_t l, double b,
    const std::vector<size_t>& s, size_t n, size_t k, size_t t) {
  size_t two_lt = Pow2(l * t);
  if (l == 0) return base_case(g, d, pred, b, s, n, k);

  auto [p, w] = find_pivots(g, d, pred, b, s, k);

  size_t m = std::max(Pow2((l - 1) * t), size_t(1));
  FrontierDS ds(m, b);
  for (size_t x : p) ds.insert(x, d[x]);

  double b0_prime = b;
  for (size_t x : p)
    if (d[x] < b0_prime) b0_prime = d[x];

  std::unordered_set<size_t> u_set;
  double last_bi_prime = b0_prime;
  size_t iter = 0;

  while (u_set.size() < k * two_lt && iter < kMaxIter) {
    double bi;
    std::vector<size_t> si;
    if (!ds.pull(bi, si)) break;
    ++iter;
    auto [bi_prime, ui] = bmssp(g, d, pred, l - 1, bi, si, n, k, t);
    last_bi_prime = bi_prime;
    for (size_t u : ui) u_set.insert(u);

    std::vector<std::pair<size_t, double>> k_list;
    for (size_t u : ui) {
      for (const auto& [v, w_e] : g.out_edges(u)) {
        double new_d = d[u] + w_e;
        if (new_d > d[v]) continue;
        Relax(d, pred, u, v, w_e);
        if (d[v] >= bi && d[v] < b)
          ds.insert(v, d[v]);
        else if (d[v] >= bi_prime && d[v] < bi)
          k_list.emplace_back(v, d[v]);
      }
    }
    for (size_t x : si)
      if (d[x] >= bi_prime && d[x] < bi) k_list.emplace_back(x, d[x]);
    ds.batch_prepend(k_list);

    if (ds.is_empty()) {
      return {b, std::vector<size_t>(u_set.begin(), u_set.end())};
    }
    if (u_set.size() > k * two_lt) {
      return {bi_prime, std::vector<size_t>(u_set.begin(), u_set.end())};
    }
  }

  double b_prime = (iter > 0) ? last_bi_prime : b0_prime;
  for (size_t x : w)
    if (d[x] < b_prime) u_set.insert(x);
  return {b_prime, std::vector<size_t>(u_set.begin(), u_set.end())};
}

}  // namespace

SsspResult duan_mao_shu_yin(const Graph& g, size_t source) {
  size_t n = g.vertex_count();
  if (n == 0) return {{}, {}};

  std::vector<double> d(n, kInf);
  std::vector<std::optional<size_t>> pred(n);
  d[source] = 0;

  double t_log = std::log(static_cast<double>(std::max(n, size_t(2)))) / std::log(2.0);
  size_t t = std::max(static_cast<size_t>(1),
                      static_cast<size_t>(std::floor(std::pow(t_log, 2.0 / 3.0))));
  size_t k = std::max(static_cast<size_t>(1),
                      static_cast<size_t>(std::floor(std::pow(t_log, 1.0 / 3.0))));
  size_t l = static_cast<size_t>(std::ceil(t_log / static_cast<double>(t)));

  std::vector<size_t> s = {source};
  bmssp(g, d, pred, l, kInf, s, n, k, t);

  return {std::move(d), std::move(pred)};
}

}  // namespace sssp

File: go/sssp.go
package sssp

import (
	"container/heap"
	"math"
	"sort"
)

const maxIter = 100_000_000
const eps = 1e-12

// SsspResult holds distances and predecessors from a single-source shortest path run.
type SsspResult struct {
	Distance    []float64
	Predecessor []int // -1 for source or unreachable
}

// VertexCount returns the number of vertices.
func (r *SsspResult) VertexCount() int {
	return len(r.Distance)
}

// dijkstraHeap is a min-heap of (distance, vertex) for Dijkstra.
type dijkstraHeap []struct {
	dist float64
	v    int
}

func (h dijkstraHeap) Len() int           { return len(h) }
func (h dijkstraHeap) Less(i, j int) bool { return h[i].dist < h[j].dist }
func (h dijkstraHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *dijkstraHeap) Push(x any)        { *h = append(*h, x.(struct{ dist float64; v int })) }
func (h *dijkstraHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

// Dijkstra runs SSSP from source using Dijkstra's algorithm (O((V+E) log V)).
// Returns the same result type as DuanMaoShuYin. Predecessor is -1 for source or unreachable.
func Dijkstra(g *Graph, source int) *SsspResult {
	n := g.VertexCount()
	if n == 0 {
		return &SsspResult{}
	}
	d := make([]float64, n)
	pred := make([]int, n)
	for i := range d {
		d[i] = math.Inf(1)
		pred[i] = -1
	}
	d[source] = 0
	cap := n
	if cap > 4096 {
		cap = 4096
	}
	pq := make(dijkstraHeap, 0, cap)
	heap.Init(&pq)
	heap.Push(&pq, struct{ dist float64; v int }{0, source})
	for pq.Len() > 0 {
		entry := heap.Pop(&pq).(struct{ dist float64; v int })
		du, u := entry.dist, entry.v
		if du > d[u] {
			continue
		}
		edges := g.OutEdges(u)
		for i := 0; i < len(edges); i++ {
			e := &edges[i]
			v, w := e.to, e.weight
			newD := d[u] + w
			if newD < d[v] {
				d[v] = newD
				pred[v] = u
				heap.Push(&pq, struct{ dist float64; v int }{newD, v})
			}
		}
	}
	return &SsspResult{Distance: d, Predecessor: pred}
}

// DuanMaoShuYin runs SSSP from source using the Duan–Mao–Shu–Yin algorithm.
func DuanMaoShuYin(g *Graph, source int) *SsspResult {
	n := g.VertexCount()
	if n == 0 {
		return &SsspResult{}
	}
	d := make([]float64, n)
	pred := make([]int, n)
	for i := range d {
		d[i] = math.Inf(1)
		pred[i] = -1
	}
	d[source] = 0

	tLog := math.Log(float64(max(n, 2))) / math.Ln2
	t := max(1, int(math.Floor(math.Pow(tLog, 2.0/3.0))))
	k := max(1, int(math.Floor(math.Pow(tLog, 1.0/3.0))))
	l := int(math.Ceil(tLog / float64(t)))

	s := []int{source}
	bmssp(g, d, pred, l, math.Inf(1), s, n, k, t)

	return &SsspResult{Distance: d, Predecessor: pred}
}

func bmssp(g *Graph, d []float64, pred []int, l int, b float64, s []int, n, k, t int) (float64, []int) {
	twoLt := pow2(l * t)
	if l == 0 {
		return baseCase(g, d, pred, b, s, n, k)
	}

	p, w := findPivots(g, d, pred, b, s, k)
	m := max(pow2((l-1)*t), 1)
	ds := newFrontierDS(m, b)
	for _, x := range p {
		ds.insert(x, d[x])
	}

	b0Prime := b
	for _, x := range p {
		if d[x] < b0Prime {
			b0Prime = d[x]
		}
	}

	uSetCap := k * twoLt
	if uSetCap > 1_000_000 {
		uSetCap = 1_000_000
	}
	uSet := make(map[int]struct{}, uSetCap)
	lastBiPrime := b0Prime
	iter := 0

	for len(uSet) < k*twoLt && iter < maxIter {
		bi, si, ok := ds.pull()
		if !ok {
			break
		}
		iter++
		biPrime, ui := bmssp(g, d, pred, l-1, bi, si, n, k, t)
		lastBiPrime = biPrime
		for _, u := range ui {
			uSet[u] = struct{}{}
		}

		kCap := 64
		if len(ui)*4 > kCap {
			kCap = len(ui) * 4
		}
		kListV := make([]int, 0, kCap)
		kListD := make([]float64, 0, kCap)
		for _, u := range ui {
			edges := g.OutEdges(u)
			for i := 0; i < len(edges); i++ {
				e := &edges[i]
				newD := d[u] + e.weight
				if newD > d[e.to] {
					continue
				}
				relax(d, pred, u, e.to, e.weight)
				if d[e.to] >= bi && d[e.to] < b {
					ds.insert(e.to, d[e.to])
				} else if d[e.to] >= biPrime && d[e.to] < bi {
					kListV = append(kListV, e.to)
					kListD = append(kListD, d[e.to])
				}
			}
		}
		for _, x := range si {
			if d[x] >= biPrime && d[x] < bi {
				kListV = append(kListV, x)
				kListD = append(kListD, d[x])
			}
		}
		ds.batchPrepend(kListV, kListD)

		if ds.isEmpty() {
			return b, mapToSlice(uSet)
		}
		if len(uSet) > k*twoLt {
			return biPrime, mapToSlice(uSet)
		}
	}

	bPrime := b0Prime
	if iter > 0 {
		bPrime = lastBiPrime
	}
	for _, x := range w {
		if d[x] < bPrime {
			uSet[x] = struct{}{}
		}
	}
	return bPrime, mapToSlice(uSet)
}

func baseCase(g *Graph, d []float64, pred []int, b float64, s []int, n, k int) (float64, []int) {
	x := s[0]
	var u0 []int
	heap := newMinHeap(n)
	heap.insert(x, d[x])

	for !heap.isEmpty() && len(u0) < k+1 {
		u, du := heap.extractMin()
		u0 = append(u0, u)
		for _, e := range g.OutEdges(u) {
			newD := du + e.weight
			if newD >= b || newD > d[e.to] {
				continue
			}
			relax(d, pred, u, e.to, e.weight)
			if heap.contains(e.to) {
				heap.decreaseKey(e.to, d[e.to])
			} else {
				heap.insert(e.to, d[e.to])
			}
		}
	}

	if len(u0) <= k {
		return b, u0
	}
	bPrime := b
	for _, v := range u0 {
		if d[v] > bPrime {
			bPrime = d[v]
		}
	}
	var filtered []int
	for _, v := range u0 {
		if d[v] < bPrime {
			filtered = append(filtered, v)
		}
	}
	return bPrime, filtered
}

func findPivots(g *Graph, d []float64, pred []int, b float64, s []int, k int) (p, w []int) {
	w = append([]int{}, s...)
	wi := append([]int{}, s...)

	for round := 1; round <= k; round++ {
		var wiNext []int
		for _, u := range wi {
			for _, e := range g.OutEdges(u) {
				newD := d[u] + e.weight
				if newD > d[e.to] {
					continue
				}
				relax(d, pred, u, e.to, e.weight)
				if newD < b {
					wiNext = append(wiNext, e.to)
				}
			}
		}
		w = append(w, wiNext...)
		wi = wiNext
		if len(w) > k*len(s) {
			return s, w
		}
	}

	inW := make(map[int]struct{}, len(w))
	for _, v := range w {
		inW[v] = struct{}{}
	}
	parent := make([]int, g.VertexCount())
	for i := range parent {
		parent[i] = -1
	}
	for _, u := range w {
		for _, e := range g.OutEdges(u) {
			if _, ok := inW[e.to]; ok && math.Abs(d[e.to]-(d[u]+e.weight)) < eps {
				if parent[e.to] == -1 {
					parent[e.to] = u
				}
			}
		}
	}

	children := make([][]int, g.VertexCount())
	for _, v := range w {
		if parent[v] >= 0 {
			children[parent[v]] = append(children[parent[v]], v)
		}
	}

	subtreeSize := make([]int, g.VertexCount())
	var countSubtree func(int) int
	countSubtree = func(u int) int {
		if subtreeSize[u] != 0 {
			return subtreeSize[u]
		}
		s := 1
		for _, v := range children[u] {
			s += countSubtree(v)
		}
		subtreeSize[u] = s
		return s
	}

	hasParent := make(map[int]struct{}, len(w))
	for _, v := range w {
		if parent[v] >= 0 {
			hasParent[v] = struct{}{}
		}
	}
	var rootsInS []int
	for _, r := range s {
		if _, ok := hasParent[r]; !ok {
			if countSubtree(r) >= k {
				rootsInS = append(rootsInS, r)
			}
		}
	}
	return rootsInS, w
}

func relax(d []float64, pred []int, u, v int, w float64) {
	newD := d[u] + w
	if newD >= d[v] {
		return
	}
	d[v] = newD
	pred[v] = u
}

func pow2(exp int) int {
	if exp <= 0 {
		return 1
	}
	if exp > 30 {
		exp = 30
	}
	return 1 << exp
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func mapToSlice(m map[int]struct{}) []int {
	out := make([]int, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	return out
}

// --- frontierDS ---
type frontierDS struct {
	m            int
	b            float64
	keyToValue   map[int]float64
	list         []struct{ v float64; k int }
	sorted       bool
}

func newFrontierDS(m int, b float64) *frontierDS {
	if m < 1 {
		m = 1
	}
	cap := m * 2
	if cap > 64000 {
		cap = 64000
	}
	return &frontierDS{
		m:          m,
		b:          b,
		keyToValue: make(map[int]float64, cap),
		list:       make([]struct{ v float64; k int }, 0, cap),
	}
}

func (f *frontierDS) insert(key int, value float64) {
	if old, ok := f.keyToValue[key]; ok && value >= old {
		return
	}
	// remove old (value, key) from list
	for i := 0; i < len(f.list); i++ {
		if f.list[i].k == key {
			f.list = append(f.list[:i], f.list[i+1:]...)
			break
		}
	}
	f.keyToValue[key] = value
	f.list = append(f.list, struct{ v float64; k int }{value, key})
	f.sorted = false
}

func (f *frontierDS) batchPrepend(keys []int, values []float64) {
	for i, key := range keys {
		value := values[i]
		if old, ok := f.keyToValue[key]; ok && value >= old {
			continue
		}
		for j := 0; j < len(f.list); j++ {
			if f.list[j].k == key {
				f.list = append(f.list[:j], f.list[j+1:]...)
				break
			}
		}
		f.keyToValue[key] = value
		f.list = append(f.list, struct{ v float64; k int }{value, key})
	}
	f.sorted = false
}

func (f *frontierDS) pull() (bound float64, keys []int, ok bool) {
	if len(f.list) == 0 {
		return f.b, nil, false
	}
	if !f.sorted {
		sort.Slice(f.list, func(i, j int) bool {
			if f.list[i].v != f.list[j].v {
				return f.list[i].v < f.list[j].v
			}
			return f.list[i].k < f.list[j].k
		})
		f.sorted = true
	}
	take := f.m
	if take > len(f.list) {
		take = len(f.list)
	}
	keys = make([]int, take)
	for i := 0; i < take; i++ {
		keys[i] = f.list[i].k
		delete(f.keyToValue, f.list[i].k)
	}
	f.list = f.list[take:]
	if len(f.list) > 0 {
		bound = f.list[0].v
	} else {
		bound = f.b
	}
	return bound, keys, true
}

func (f *frontierDS) isEmpty() bool {
	return len(f.list) == 0
}

// --- minHeap ---
type minHeap struct {
	heap  []struct{ v int; d float64 }
	index []int
}

func newMinHeap(maxVertex int) *minHeap {
	idx := make([]int, maxVertex)
	for i := range idx {
		idx[i] = -1
	}
	heapCap := maxVertex
	if heapCap > 4096 {
		heapCap = 4096
	}
	return &minHeap{
		heap:  make([]struct{ v int; d float64 }, 0, heapCap),
		index: idx,
	}
}

func (h *minHeap) isEmpty() bool {
	return len(h.heap) == 0
}

func (h *minHeap) contains(v int) bool {
	return v < len(h.index) && h.index[v] >= 0
}

func (h *minHeap) insert(v int, d float64) {
	i := len(h.heap)
	h.heap = append(h.heap, struct{ v int; d float64 }{v, d})
	h.index[v] = i
	h.siftUp(i)
}

func (h *minHeap) extractMin() (int, float64) {
	top := h.heap[0]
	h.index[top.v] = -1
	h.heap[0] = h.heap[len(h.heap)-1]
	h.heap = h.heap[:len(h.heap)-1]
	if len(h.heap) > 0 {
		h.index[h.heap[0].v] = 0
		h.siftDown(0)
	}
	return top.v, top.d
}

func (h *minHeap) decreaseKey(v int, newD float64) {
	i := h.index[v]
	if i < 0 || h.heap[i].d <= newD {
		return
	}
	h.heap[i] = struct{ v int; d float64 }{v, newD}
	h.siftUp(i)
}

func (h *minHeap) siftUp(i int) {
	for i > 0 {
		p := (i - 1) / 2
		if h.heap[p].d <= h.heap[i].d {
			break
		}
		h.swap(i, p)
		i = p
	}
}

func (h *minHeap) siftDown(i int) {
	for {
		l, r, small := 2*i+1, 2*i+2, i
		if l < len(h.heap) && h.heap[l].d < h.heap[small].d {
			small = l
		}
		if r < len(h.heap) && h.heap[r].d < h.heap[small].d {
			small = r
		}
		if small == i {
			break
		}
		h.swap(i, small)
		i = small
	}
}

func (h *minHeap) swap(i, j int) {
	a, b := h.heap[i], h.heap[j]
	h.heap[i], h.heap[j] = b, a
	h.index[a.v], h.index[b.v] = j, i
}
